### PR TITLE
Add `--system-root` parameter, `systemRoot()` and `path()` functions  to dsc

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

To enable offline configuration management, add `--system-root` parameter and `systemRoot()` function (takes no arguments) to DSC.  The provided path must exist.  Supported resources would need to accept the path to use rather than the current running OS.
If `--system-root` is not specified, then `systemRoot()` will default to `$env:SYSTEMDRIVE` on Windows and `/` on Linux/macOS.
Also added `path()` function that takes any number of strings and concatenates it with the OS directory separator char.

## PR Context

Fix https://github.com/PowerShell/DSC/issues/534

### CODE REVIEW

File: dsc_lib/src/functions/mounted_path.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+use crate::DscError;
+use crate::configure::context::Context;
+use crate::functions::{AcceptedArgKind, Function};
+use serde_json::Value;
+use tracing::debug;
+
+#[derive(Debug, Default)]
+pub struct MountedPath {}
+
+/// Implements the 'mountedpath' function.
```
Comment (michaeltlombardi) [inline dsc_lib/src/functions/mounted_path.rs:13]: Should this use camel casing, like `createArray` and `resourceId`?

File: dsc_lib/src/functions/mod.rs [hunk 2]
```diff
@@ -74,6 +75,7 @@ impl FunctionDispatcher {
         functions.insert("max".to_string(), Box::new(max::Max{}));
         functions.insert("min".to_string(), Box::new(min::Min{}));
         functions.insert("mod".to_string(), Box::new(mod_function::Mod{}));
+        functions.insert("mountedpath".to_string(), Box::new(mounted_path::MountedPath{}));
```
Comment (michaeltlombardi) [inline dsc_lib/src/functions/mod.rs:78]: Should this be camel cased, like `createArray` and `resourceId`?

File: dsc_lib/src/configure/context.rs [hunk 3]
```diff
@@ -12,6 +12,7 @@ use super::config_doc::{DataType, SecurityContextKind};
 pub struct Context {
     pub execution_type: ExecutionKind,
     pub outputs: HashMap<String, Value>, // this is used by the `reference()` function to retrieve output
+    pub mounted_path: String,
```
Comment (tgauth) [inline dsc_lib/src/configure/context.rs:15]: How did you decide on using a `String`? Curious if `PathBuf` would be useful here.

Comment (SteveL-MSFT) [reply dsc_lib/src/configure/context.rs:15]: Should probably be PathBuf as appending takes care of dir separator automatically.

File: dsc_lib/src/functions/target_path.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+use crate::DscError;
+use crate::configure::context::Context;
+use crate::functions::{AcceptedArgKind, Function};
+use serde_json::Value;
+use tracing::debug;
+
+#[derive(Debug, Default)]
+pub struct TargetPath {}
+
+/// Implements the 'mountedpath' function.
```
Comment (tgauth) [inline dsc_lib/src/functions/target_path.rs:13]:
```suggestion
/// Implements the 'TargetPath' function.
```
there's a few other references to mounted path in this file and others

**Comment by tgauth** [comment]
Would it be helpful to have some sort of `info` logging (maybe in the parser?) in the case of `--target-path` being provided without `TargetPath()` or is it obvious that `--target-path` would be ignored?

**Comment by SteveL-MSFT** [comment]
> Would it be helpful to have some sort of `info` logging (maybe in the parser?) in the case of `--target-path` being provided without `TargetPath()` or is it obvious that `--target-path` would be ignored?

That could be useful as a warning message if the user provided `--target-path` but the configuration doesn't use it.  The challenge is that expressions are parsed at runtime, so the warning would only happen at the end after configuration was run.  I think this is ok.

**Comment by SteveL-MSFT** [comment]
Decided to rename to `systemRoot` instead of `targetPath` as it more accurately describes what it does

**Review by tgauth — APPROVED** [review]

**Pull request merged**